More system calls

skuenzer opened the pull request:
Registers a number of system calls. Work was done towards EuroSys'21 submission by Constantin Raducanu and myself. This series depends on PR #172.

skuenzer committed:
lib/uksysinfo: Use Linux ABI numbers for `sysinfo` attribute

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

consra committed:
lib/vfscore: Register `dup` to syscall_shim

Registers `dup` system call to syscall_shim library.

Signed-off-by: Constantin Raducanu <raducanu.costi@gmail.com>
Checkpatch-Ignore: INDENTED_LABEL

consra committed:
lib/vfscore: Add the macro UK_SYSCALL_R_DEFINE for dup

Adds the macro UK_SYSCALL_R_DEFINE for `dup` in order to
be registered to syscall_shim library.

Signed-off-by: Constantin Raducanu <raducanu.costi@gmail.com>

consra committed:
lib/vfscore: Register `open` to syscall_shim

Registers `open` system call to syscall_shim library.
The signature of the `open` library call is different from signature
of the `open` syscall, forcing us to use the macro UK_LLSYSCALL_R_DEFINE.

Signed-off-by: Constantin Raducanu <raducanu.costi@gmail.com>
Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>
Checkpatch-Ignore: INDENTED_LABEL

consra committed:
lib/vfscore: Register `pipe` to syscall_shim

Registers `pipe` system call to syscall_shim library.

Signed-off-by: Constantin Raducanu <raducanu.costi@gmail.com>

skuenzer committed:
lib/vfscore: Register `openat` to syscall_shim

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/posix-process: Initial `prlimit64`

Provides getrlimit(), setrlimit(), and prlimit() with prlimit64 system
call. This initial implementation supports retrieving size of the
stack and number of max open files. Setting of values is ignored.

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/vfscore: Register `fcntl` to syscall_shim

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/vfscore: fcntl(): Warn stubbed functions only once

Warn only once when stubbed functions are called: `F_SETLK`,
`F_GETLK`, `F_SETLKW`, and `F_SETOWN`

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/syscall_shim: Print system call handler on debug

When debugging of the library (or globally) is enabled a debug message
is print whenever a registered system call handler is called.

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>
Checkpatch-Ignore: SPACING
Checkpatch-Ignore: COMPLEX_MACRO

skuenzer committed:
lib/vfscore: Register 'pread64' to syscall_shim

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/vfscore: Register 'pwrite64' to syscall_shim

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/posix-process: Register fork, vfork to syscall_shim

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/posix-process: Register execve to syscall_shim

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/posix-process: Register wait4 to syscall_shim

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/posix-process: Register 'getrlimit', 'setrlimit' to syscall_shim

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/posix-process: Register 'prctl' to syscall_shim

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/ukswrand: Register `getrandom` to syscall_shim

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/posix-sysinfo: Register 'sysinfo' to syscall_shim

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/uksched: Provide & register 'sched_yield' to syscall_shim

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/uktime: Register 'clock_settime' to syscall_shim

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/uktime: Register 'clock_getres' to syscall_shim

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/uktime: Register 'setitimer' to syscall_shim

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/uktime: Register timer functions to syscall_shim

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/uktime: Remove 'utime'

'utime' is actually VFS-related. It should get introduced with 'vfscore'.

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/vfscore: Register 'utime' to syscall_shim

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

razvand commented:
@skuenzer , could you please fix the checkpatch reports and rebase the PR before I will take a swing at it? I've reviewed PR #172 , it's been automatically merged.

mogasergiu commented on lib/vfscore/exportsyms.uk:
```diff
@@ -81,16 +85,23 @@ uk_syscall_e_lstat
 readdir_r
 readdir64
 closedir
 pread
```
Shouldn't there also be a `pread64` added (similarly to how `pwrite64` is registered)?

mogasergiu commented on lib/vfscore/main.c:
```diff
@@ -329,24 +348,44 @@ UK_TRACEPOINT(trace_vfs_pread, "%d %p 0x%x 0x%x", int, void*, size_t, off_t);
 }
 
+#if UK_LIBC_SYSCALLS
+ssize_t pread(int fd, void *buf, size_t count, off_t offset)
```
This does bring the libc-style wrapper for `pread` but would cause an undefined reference if the libc-style wrapper of `pread64` were to be used.

skuenzer replied:
Hum... I was actually expecting that the libc is mapping `pread64()` to `pread()` with a macro, so the other libc wrapper would not be needed. In fact, on an libc implementation that does not do this the symbol would be missing. I guess it does not hurt if I provide both symbols: The libc implementation makes the decision if we link to both or just one of them and the other symbol would be unused and could be removed.

skuenzer replied:
Hey @mogasergiu, I enabled the 64 variants of the system call symbols with the already available `LFS64()` macro. It is a separate commit on the updated PR. Thanks, Simon

mogasergiu commented on lib/vfscore/main.c:
```diff
@@ -498,24 +537,44 @@ UK_TRACEPOINT(trace_vfs_pwrite, "%d %p 0x%x 0x%x", int, const void*, size_t,
 }
 
+#if UK_LIBC_SYSCALLS
+ssize_t pwrite(int fd, const void *buf, size_t count, off_t offset)
```
This does bring the libc-style wrapper for `pwrite` but would cause an undefined reference if the libc-style wrapper of `pwrite64` were to be used.

skuenzer replied:
This is the same as the `pread` argument. I will provide both `64` versions of the symbols.

mogasergiu commented on lib/posix-process/process.c:
```diff
@@ -342,8 +344,109 @@ UK_SYSCALL_R_DEFINE(int, setpriority, int, which, id_t, who, int, prio)
+		    (unsigned long long) old_limit->rlim_max);
+	return 0;
+}
+
```
These `getrlimit`, `setrlimit` and `getrusage` seem to cause a build error with newlib due to what I think to be a conflict with the ones already [there](https://github.com/unikraft/lib-newlib/blob/9a7aa73ba7526e88c5c9636667e3e012f2042c81/resource.c#L41).

skuenzer replied:
True, the fix is to remove the implementation from `newlib`. I will send out an separate PR for this.

mogasergiu reviewed (changes requested):
Besides the previous comments and the checkpatch errors, everything seems to work fine.

Reviewed-by: Sergiu Moga <sergiu.moga@protonmail.com>

nderjung reviewed (dismissed):
Please rebase before merge. :)

skuenzer committed:
lib/vfscore: 64-bit symbol aliases with `LFS64()`

This commit enables the 64-bit variant of system call symbols that
are marked with the `LFS64()` macro. Such symbols are VFS system
calls, like `open()`, `openat()`, `fstafs()`, `pwritev()` where
historically a 64 bit variant was introduced. Since our implementation
already targets for 64-bit file addressing, we introduce the 64-bit
symbols as aliases.

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
include/essentials: Provide `__alias()`

Similar to `__weak_alias()` we introduce a macro for
creating a strong symbol alias (instead of weak).

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>
Checkpatch-Ignore: COMPLEX_MACRO

mogasergiu reviewed (changes requested):
Everything should be fine now. All that is left is bringing the corrections indicated by checkpatch and the separate PR for newlib.

Reviewed-by: Sergiu Moga <sergiu.moga@protonmail.com>

nderjung commented:
Hey @mogasergiu, to get this merged, please change your review to "approved" with the signoff, thanks! :)

mogasergiu commented on lib/uksched/sched.c:
```diff
@@ -286,3 +287,9 @@ void uk_sched_thread_exit(void)
 	UK_CRASH("Failed to stop the thread\n");
 }
+
+UK_SYSCALL_R_DEFINE(int, sched_yield)
```
Besides the `newlib` conflict previously mentioned, there also seems to be a conflict when compiling with `pthread-embedded`. To be more precise, it conflicts with the declaration located in the [extracted source code](https://github.com/unikraft/lib-pthread-embedded/blob/c59d7d632c9fc50f8ebe4e24139ff01265f14f9c/Makefile.uk#L189)

skuenzer commented:
I just noticed this checkpatch error:
```
ERROR:BAD_SIGN_OFF: Unrecognized email address: 'Simon Kuenzer <simon.kuenzer@neclab.eu'
#9: 
Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu
```

It should be fixed now. I hope our CI/CD is properly picking everything up. 🤞🏻

skuenzer commented:
It still did not pick-up my Checkpatch-Ignores: 😢 
```
Checkpatch-Ignore: SPACING
Checkpatch-Ignore: COMPLEX_MACRO
```

mogasergiu reviewed (approved):
Looks good to me.

Reviewed-by: Sergiu Moga <sergiu.moga@protonmail.com>
